network/server.go

```diff
@@ -349,11 +358,27 @@ func (s *Server) numPeers() int64 {
 }
 
 func (s *Server) getRandomBootNode() *peer.AddrInfo {
-	randNum, _ := rand.Int(rand.Reader, big.NewInt(int64(len(s.discovery.bootnodes))))
+	if size := int64(len(s.discovery.bootnodes)); size > 0 {
+		randNum, _ := rand.Int(rand.Reader, big.NewInt(size))
+
+		return s.discovery.bootnodes[randNum.Int64()]
+	}
 
-	return s.discovery.bootnodes[randNum.Int64()]
+	return nil
 }
 
+// getBootNode returns the address of a random bootnode which is not connected
+func (s *Server) getBootNode() *peer.AddrInfo {
+	size := int64(len(s.discovery.bootnodes))
+	for i := 0; i < int(size); i++ {
+		randNum, _ := rand.Int(rand.Reader, big.NewInt(size))
+		if peer := s.discovery.bootnodes[randNum.Int64()]; !s.hasPeer(peer.ID) {
+			return peer
+		}
+	}
+
+	return nil
+}
```

Comment:
I'm not sure if this function works. Is there a possibility of that `rand.Int` returns always the same value during loop. (Of course, it's rare but it's possible)

You might want to create slice of peers that don't connect to the node first.

Reply:
Modified it, Commit [link](https://github.com/0xPolygon/polygon-edge/pull/343/commits/b798474b4de8a2439a053b5918ef98f0dc95ed48)